core/new-gui/src/app/dashboard/user/component/user-dataset/user-dataset-explorer/user-dataset-file-renderer/user-dataset-file-renderer.component.ts

```diff
@@ -0,0 +1,307 @@
+  [MIME_TYPES.JSON]: 1 * 1024 * 1024, // 1 MB for JSON files
+  [MIME_TYPES.MSEXCEL]: 10 * 1024 * 1024, // 10 MB for Excel spreadsheets
+  [MIME_TYPES.MP4]: 50 * 1024 * 1024, // 50 MB for MP4 videos
+  [MIME_TYPES.MP3]: 10 * 1024 * 1024, // 10 MB for MP3 audio files
+  [MIME_TYPES.OCTET_STREAM]: 5 * 1024 * 1024, // Default size for other binary formats
+};
+
+@Component({
+  selector: "texera-user-dataset-file-renderer",
+  templateUrl: "./user-dataset-file-renderer.component.html",
+  styleUrls: ["./user-dataset-file-renderer.component.scss"],
+})
+export class UserDatasetFileRendererComponent implements OnInit, OnChanges {
+  private DEFAULT_MAX_SIZE = 5 * 1024 * 1024; // 5 MB
+
+  public fileURL: string | undefined;
+  // safe url is used to display some formats including image
+  public safeFileURL: SafeUrl | undefined;
+
+  // table related control
+  public displayCSV: boolean = false;
+  public displayXlsx: boolean = false;
+  public tableDataHeader: any[] = [];
+  public tableContent: any[][] = [];
+  // pagination for table-like data
+  currentPageIndex = 0;
+  pageSizeOptions = [3, 5, 10, 20, 50];
+  pageSize = 20;
+
+  // image related control
+  public displayImage: boolean = false;
+
+  // markdown control
+  public displayMarkdown: boolean = false;
+
+  // json control
+  public displayJson: boolean = false;
+
+  // video
+  public displayMP4: boolean = false;
+
+  // audio
+  public displayMP3: boolean = false;
+
+  // plain text & octet stream related control
+  public displayPlainText: boolean = false;
+  public textContent: string = "";
+
+  // control flags
+  public isLoading: boolean = false;
+  public isFileSizeUnloadable = false;
+  public isFileLoadingError: boolean = false;
+  public isFileTypePreviewUnsupported: boolean = false;
+
+  public currentFile: File | undefined = undefined;
+  @Input()
+  isMaximized: boolean = false;
+
+  @Input()
+  did: number | undefined;
+
+  @Input()
+  dvid: number | undefined;
+
+  @Input()
+  filePath: string = "";
+
+  @Output()
+  loadFile = new EventEmitter<{ file: string; prefix: string }>();
+
+  constructor(
+    private datasetService: DatasetService,
+    private sanitizer: DomSanitizer,
+    private notificationService: NotificationService
+  ) {}
+
+  ngOnInit(): void {
+    this.reloadFileContent();
+  }
+
+  ngOnChanges(changes: SimpleChanges): void {
+    if (changes.did || changes.dvid || changes.filePath) {
+      this.reloadFileContent();
+    }
+  }
+
+  ngOnDestroy(): void {
+    if (this.fileURL) {
+      URL.revokeObjectURL(this.fileURL);
+    }
+  }
+
+  showImageModal = false;
+
+  toggleImageModal() {
+    this.showImageModal = !this.showImageModal;
+  }
+
+  reloadFileContent() {
+    this.turnOffAllDisplay();
+    this.isLoading = true;
+    if (this.did && this.dvid && this.filePath != "") {
+      this.datasetService
+        .retrieveDatasetVersionSingleFile(this.did, this.dvid, this.filePath)
+        .pipe()
```

Comment:
Use `.pipe(untilDestroyed(this))` instead of empty `pipe()`.

Reply:
Done